### PR TITLE
[Example] Add tnd_shared_prefix_fa: fused shared-prefix GQA FlashAttention on deduplicated TND layout

### PR DESCRIPTION
**Pull request by zwj-sep**
## 背景

搜索相关性场景中，同一 query 的多个候选文档共享完全相同的 query 前缀。直接打包时前缀逐份重复：

```
原生 TND: [prefix, a0..an, prefix, b0..bn, prefix, c0..cn]
去重 TND: [prefix, a0..an, b0..bn, c0..cn]
```

前缀去重对 element-wise 算子无影响，但 **attention 的 QK^T 语义要求每个私有段的 Q 同时看到 prefix KV 与本段私有 KV**，普通 FA 算子无法处理 KV 分布在两处的去重布局。业界常规的两步方案（先算 prefix KV 并缓存、再拼接计算）在此业务下序列短，缓存构建、KV concat 物化与多次 launch 的固定开销占比过高。

本 PR 新增 **单 kernel 融合算子** `tnd_shared_prefix_fa`：输入/输出均为去重 TND 格式 `[token, head, dim]`，边界为投影后的 Q/K/V（投影不在算子内），支持 GQA、causal mask、变长序列与非整除尾块。

### vs PyTorch (CANN) 基线（`perf_3way.py`）

基线为无融合算子时的自然做法：逐 request `torch.cat` KV + 布局转 BNSD + `npu_fusion_attention` 循环（已对照 golden 验证，max_abs 2.4e-4）。

| 场景 | Kernel-only | E2E | PyTorch(CANN) | Kernel 加速 | E2E 加速 |
|------|:---:|:---:|:---:|:---:|:---:|
| small (batch=2) | 0.12 ms | 0.35 ms | 0.29 ms | **2.43x** | 0.81x |
| business (batch=10) | 1.39 ms | 1.70 ms | 1.69 ms | **1.21x** | 1.00x |
| p99 (batch=10, seq=218) | 1.41 ms | 1.68 ms | 1.70 ms | **1.21x** | 1.01x |

> E2E 含 `block_metadata` 构造（CPU + H2D 约 0.3ms 固定开销），小 batch 下占比高；kernel 本身消除逐 request concat/布局转换/多 launch 开销。

### Developer vs Expert

| | Developer 版 | Expert 版 |
|---|:---:|:---:|
| Latency (do_bench) | 3.95 ms | **1.39 ms（2.84x）** |
| Kernel duration (msprof) | 4.25 ms | 1.38 ms |
| L0 精度（严格阈值） | 14/14 PASS | 5/5 PASS |

配置：batch=10, q_head=14, kv_head=2, head_dim=64, sp_len=24, priv_len=[150]×10, fp16。

两个版本是同一算子的两条实现路径（与 design.md 的升级路径一致）：Developer 版作为基线，Expert 版追求性能。

## 算法

在线 softmax 分两段迭代 KV：先迭代 shared 段（全部 request 共享一份 K/V），再迭代本 task 的 private 段，`m_i / sumexp / acc_o` 状态跨段累积不重置。变长通过 `block_metadata` 查表实现：编译期用 `max_private_kv_len` 作为循环上界（`T.serial` 需要编译期边界），运行时按元数据跳过空迭代并 mask。

非整除场景三类 mask：空迭代整块 `fill(NEG_INF)`、KV 尾块逐元素 `NEG_INF`、Q 尾块限制有效输出行。

## Developer 版（基线）

标准 Developer 模式：`alloc_shared/alloc_fragment` + `threads=2` + 4 个 pass_configs，在线 softmax 复用 `flash_attn_bshd_developer.py` 结构。

两个工程 workaround（均已定位根因）：
1. **CombineCV sync mismatch**：if 分支包裹 GEMM 会导致 Cube/Vector sync point 不对齐（Vec 侧 `copy_gm_to_ub` 被丢弃）。改为 GEMM + softmax 无条件执行、只分支 `T.copy`，空迭代结果由 mask 丢弃；
2. **条件 fill 被重排覆盖**：`T.tile.fill(NEG_INF)` 与 workspace 读同 buffer，编译器将其重排到 `copy_gm_to_ub` 之前导致失效。放到 `T.tile.add` 之后、`T.tile.mul` 之前，靠依赖链锁定顺序。

覆盖 **L0×5 + L1×6 + L2×1 + Boundary×2 = 14/14**，含 bf16、causal、变长混合、MHA、空 private 段、无 shared prefix、GQA 约束异常等。

## Expert 版（性能）

参照 `fa_opt/flash_attn_bhsd_expert_h16_d128.py` 架构适配 TND 布局：
- **persistent 调度**：24 核静态均分 task（`my_start/my_count`），跨核 flag 限定在核组内
- **6 跨核 flag**（S/P/O 各一对 C2V/V2C）+ `cross_interval=2` 批量同步 + `num_stages` workspace 维度
- **双缓冲 L0A/L0B/L0C** + `T.mma`
- 两段 KV 合并为单循环，按 `idx < max_shared_iters` 切换 K/V 来源

调试过程中定位并解决：
- **persistent 下逐 tile 同步死锁**：跨核 flag 的 set/wait 必须配对在批量边界（参考实现的 `cross_interval` 模式），逐 tile 同步在多核组下 flag 计数错位；
- **Q 尾块 vid=1 垃圾覆盖**：`q_valid < half_M` 时 vid=1 的输出行覆盖相邻 tile 数据，改为逐行条件写；
- **score 路径 fp16 舍入**：workspace_s/o 用 fp32 消除 L0C→GM→UB 路径的精度损失（workspace_p 保持 fp16：Cube 侧 `copy_gm_to_l1` 要求类型一致），严格阈值下 5/5 通过且性能无损。

## 文件

```
examples/tnd_shared_prefix_fa/
├── tnd_shared_prefix_fa_developer.py   # Developer 版 + L0/L1/L2/Boundary 测试（自包含）
├── tnd_shared_prefix_fa_expert.py      # Expert 版 + L0 测试（自包含）
├── perf_3way.py                         # 3-way 性能基线（kernel/E2E/PyTorch CANN）
├── design.md                            # 设计文档
└── proto.yaml                           # 接口规格（覆盖门禁用）
```

## 测试

```bash
# Developer 版（L0/L1/L2/Boundary 14 用例）
python examples/tnd_shared_prefix_fa/tnd_shared_prefix_fa_developer.py --level all

# Expert 版（L0 5 用例）
python examples/tnd_shared_prefix_fa/tnd_shared_prefix_fa_expert.py --level l0

# 3-way 性能基线
python examples/tnd_shared_prefix_fa/perf_3way.py
```

### msprof kernel 级分析（可选）

用任意脚本包住 kernel 调用后由 msprof 启动，例如：

```python
# profile_run.py: 构造输入 -> tnd_shared_prefix_fa_expert(...) -> 循环跑若干次
```

```bash
msprof --output=/tmp/msprof_tnd --application="python profile_run.py"
```

### CODE REVIEW

File: examples/tnd_shared_prefix_fa/tnd_shared_prefix_fa_developer.py [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+    total_private_kv,
+    total_q_blocks,
+    block_M=128,
+    block_N=64,
+    sm_scale=None,
+    dtype_str="float16",
+    causal_mask=False,
+    threads=2,
+):
+    sm_scale = (1.0 / head_dim) ** 0.5 if sm_scale is None else sm_scale
```
Comment (fengz72) [inline examples/tnd_shared_prefix_fa/tnd_shared_prefix_fa_developer.py]: The sm_scale expression is mathematically equivalent to $\frac{1}{\sqrt{\text{head\\_dim}}}$, but its form—divide first, then square-root—is unconventional and less readable. The standard form is clearer:
`sm_scale = 1.0 / (head_dim ** 0.5)`

File: examples/tnd_shared_prefix_fa/tnd_shared_prefix_fa_developer.py [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+    dtype = dtype_str
+    accum_dtype = "float32"
+    group_size = q_head // kv_head
```
Comment (fengz72) [inline examples/tnd_shared_prefix_fa/tnd_shared_prefix_fa_developer.py:51]: assert q_head % kv_head == 0

File: examples/tnd_shared_prefix_fa/tnd_shared_prefix_fa_developer.py [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+
+
+@tilelang.jit(out_idx=[6], pass_configs=pass_configs)
+def tnd_shared_prefix_fa_developer(
+    q_head,
+    kv_head,
+    head_dim,
+    shared_prefix_len,
+    max_private_kv_len,
+    total_q,
```
Comment (fengz72) [inline examples/tnd_shared_prefix_fa/tnd_shared_prefix_fa_developer.py:37]: total_q / shared_prefix_len / max_private_kv_len / total_private_kv / total_q_blocks are passed as Python ints into T.Tensor([...]), so any value change forces recompilation. Converting them to T.symbolic makes these dimensions symbolic—compile once, pass values at runtime.

**Review by fengz72 — COMMENTED** [review]
Recommend providing performance data.

**Comment by zwj-sep** [comment]
Thanks for the review @fengz72! All three issues addressed in commit 147693a:

1. **sm_scale**: Changed to `1.0 / (head_dim**0.5)` for clarity.
2. **GQA assert**: Added `assert q_head % kv_head == 0`.
3. **T.symbolic**: Converted `total_q`, `total_private_kv`, `shared_prefix_len`, `total_q_blocks` to `T.symbolic` in tensor shapes — compile once, pass values at runtime. `shared_prefix_len` and `max_private_kv_len` remain Python ints because they determine `T.serial` loop bounds (compile-time), as noted in design.md §10.1.

**Review by fengz72 — COMMENTED** [review]
搜索相关性场景中，同一 query 的多个候选文档共享完全相同的 query 前缀：
```text
原生 TND 打包：[prefix, a0..an, prefix, b0..bn, prefix, c0..cn]
去重后：      [prefix, a0..an, b0..bn, c0..cn]
```

- 为何需要定制算子：前缀逐 token 相同，去重对 element-wise 算子无影响；但 attention 中每个私有段的 Q 必须同时看到 prefix KV 与本段私有 KV（QK^T 语义所迫），故需新算子支持
- 为何不用两步方案：大模型 shared-prefix 常规做法是 
    - step1 单算 prefix 并缓存 K/V_prefix，
    - step2 私有 token 计算时拼接缓存。
    - 此业务请求短，缓存构建、KV concat 物化与多次 launch 的固定开销占比过高，故选择单 kernel 融合
- 输入/输出均为去重 TND 格式
- 算子边界：输入是 QKV，即投影后的 Q/K/V，形状一致；投影不在算子内
